### DISCUSSION TITLE
Changing score pitch.

### DISCUSSION THREAD

**Discussion by matevzsenlab** [CoderLine/alphaTab#1218] [Q&A] [answered]
Hi. I was wondering if it's possible to display the same notes in the score but change the pitch of an entire score. I found that there are displayTranspositionPitch and transpositionPitch options but it's not exactly what I'm looking for. I also tried combining these two so I would use transpositionPitch to pitch song +2 semitones and then use displayTranspositionPitch to display the original notes, but displayTranspositionPitch only affects standard music notation and not tab notation as well.

**Comment by Danielku15** [comment]
* the displayTranspositionPitch only changes the standard notation. This comes from the wierd rule that Guitar notation is transposed by 1 octave without any visual indication (like a different clef or ottava - as how it is indicated normally). It will not have an effect on the audio. 
* the transpositionPitch is expected to affect the standard notation and playback [by changing the "value" of the notes](https://github.com/CoderLine/alphaTab/blob/de959d4c6b9a736cf4e5c8483fc3f9d76800fa83/src/model/Note.ts#L506-L529) and the guitar tablatures [in a very simple way of shifting frets](https://github.com/CoderLine/alphaTab/blob/de959d4c6b9a736cf4e5c8483fc3f9d76800fa83/src/rendering/glyphs/NoteNumberGlyph.ts#L29-L32) 

Technically it should also be possible to even combine the settings. e.g. have -12 on the `displayTranspositionPitch` for a guitar track to pull it down one octave, and then put a `transpositionPitch` of -2 on top to transpose it to be played lower. 

Unfortunately we don't have yet a more advanced transposition which would still attempt to make it nicely playable on a guitar. e.g. you might end up with negative frets.

Can you maybe setup a fiddle showing your setup and what doesn't work exactly as you need it? Could be I understand something wrong or something is broken.

**Reply by matevzsenlab** [reply]
Thank you for the answer.

No, you understand correctly. I just wanted to keep the notes on the standard and tab notation the same and change the pitch. That's why I was asking if it is possible to achieve that with combination of displayTranspositionPitch and transpositionPitch. 

If you want an example, you can go to any Ultimate guitar interactive tabs (any Guitar pro tab) and try to change pitch in the controls. It's a really useful feature for learning songs without changing your guitar's tuning.

**Reply by Danielku15** [reply]
You can achieve this by only using `transpositionPitch` and not modifying `displayTranspositionPitch`. 

https://jsfiddle.net/danielku15/kx0Lqhbo/1/

But you have to consider this open work item: https://github.com/CoderLine/alphaTab/issues/896

Including workaround for playback: 
https://jsfiddle.net/danielku15/rbw8c2h0/1/

**Reply by matevzsenlab** [reply]
Yes, I understand how these two properties work. I am trying to change the pitch of the playback audio and keep the displayed tabs and notations as they are. Like I said, the best example is on Ultimate guitar. If you go on any guitar pro tab on their site, you can do it like this:
![chrome_Mc4wgI93Li](https://github.com/CoderLine/alphaTab/assets/101171504/d26797fc-5260-45f4-90e0-0b70e53ed8b1)

![chrome_6zKrcq835o](https://github.com/CoderLine/alphaTab/assets/101171504/ee29fe5d-fad3-4a89-ae8a-1d20e60dbc78)

This is really helpful when you try to play the song, without having to tune to the actual tuning of the song. You can just change the pitch of the entire score, and you can then easily play along.

**Reply by Danielku15** [reply, marked as answer]
I added a future work item here: https://github.com/orgs/CoderLine/projects/12/views/1?pane=issue&itemId=38121145

**Comment by GaryFrewin** [comment]
@Danielku15  Not sure if this is related. I was trying to create a player that loops and ascends in pitch by a semitone each time... 

Simple script:

```
    <script type="text/javascript">
  
        const settings = {
            file: "Rossini.gp",
            display: {
                layoutMode: "horizontal"
            },
            notation: {
                elements: {
                    scoreTitle: false,
                    scoreSubTitle: false,
                    trackNames: false,
                    scoreArtist: false, 
                    guitarTuning: false,
                    effectTempo: false
                }
            },
            player: {
                enablePlayer: true,
                soundFont: 'https://cdn.jsdelivr.net/npm/@coderline/alphatab@latest/dist/soundfont/sonivox.sf2',
                scrollElement: wrapper.querySelector('.at-viewport') // this is the element to scroll during playback
            }
        };

        const api = new alphaTab.AlphaTabApi(main, settings);
        
        // This works: The pitch is higher on first play
        api.scoreLoaded.on((score) => {
            score.tracks[0].staves[0].transpositionPitch = -2
            score.tracks[0].staves[0].displayTranspositionPitch = -2
        });

        // This changes it visually, but it doesn't change audibly. 
        api.playerFinished.on((args) => {
            api.stop();
            api.score.tracks[0].staves[0].transpositionPitch = -4
            api.score.tracks[0].staves[0].displayTranspositionPitch = -2

            
            api.updateSettings();
            api.render();
            api.playPause();
        });
```

Am I missing something here? 

Thanks in advance for any guidance.

**Reply by Danielku15** [reply]
Yes. Looks related partially. Changing the two options currently doesn't regenerate the Midi used internally for playback. This is tracked here: 

https://github.com/CoderLine/alphaTab/issues/896

You could try calling `api.loadMidiForScore()`. It is not an official and exposed API but I think it should be possible to call it as we are not hiding the private members: 

https://github.com/CoderLine/alphaTab/blob/develop/src/AlphaTabApiBase.ts#L602
